sycl/doc/KernelProgramCache.md

```diff
@@ -68,44 +69,113 @@ times, which may involve JIT compilation and take quite a lot of time.
+ - build options id <sup>[2](#what-is-bopts)</sup>.
 
-The krnels map's key consists of three components too: program the kernel
-belongs to, kernel name<sup>[2](#what-is-kname)</sup>, device the program is
-built for.
+The kernels map's key consists of two components:
+ - program the kernel, belongs to,
+ - kernel name<sup>[3](#what-is-kname)</sup>.
 
 <a name="what-is-ksid">1</a>: Kernel set id is an ordinal number of the device
 binary image the kernel is contained in.
 
-<a name="what-is-kname">2</a>: Kernel name is a kernel ID mangled class' name
-which is provided to methods of `cl::sycl::handler` (e.g. `parallel_for` or
-`single_task`).
+<a name="what-is-bopts">2</a>: The concatenation of build options (both compile
+and link options) set in application or environment variables. There are three
+sources of build options which cache is aware of:
+ - from device image (pi_device_binary_struct::CompileOptions,
+ pi_device_binary_struct::LinkOptions);
+ - environment variables (SYCL_PROGRAM_COMPILE_OPTIONS,
+ SYCL_PROGRAM_LINK_OPTIONS);
+ - options passed through SYCL API.
 
+Note: Backend runtimes used by SYCL can have extra environment or configurations
+values (e.g. IGC has [igc_flags.def](https://github.com/intel/intel-graphics-compiler/blob/7f91dd6b9f2ca9c1a8ffddd04fa86461311c4271/IGC/common/igc_flags.def) which affect JIT process). Changing such
+configuration will invalidate cache and manual cache cleanup should be done.
 
-## Points of improvement (things to do)
-
- - Implement LRU policy on cached objects. See [issue](https://github.com/intel/llvm/issues/2517).
- - Allow for caching of objects built with some build options.
- - Employ the same built object for multiple devices of the same ISA,
-   capabilities and so on. *NOTE:* It's not really known if it's possible to
-   check if two distinct devices are *exactly* the same. Probably this should be
-   an improvement request for plugins.
- - Improve testing: cover real use-cases. See currently covered cases [here](https://github.com/intel/llvm/blob/sycl/sycl/unittests/kernel-and-program/Cache.cpp).
+<a name="what-is-kname">3</a>: Kernel name is a kernel ID mangled class' name
+which is provided to methods of `cl::sycl::handler` (e.g. `parallel_for` or
+`single_task`).
 
+### Persistent cache
+
+The cache works behind in-memory cache and stores the same underlying PI
+object behind `cl::sycl::program` user-level objects in a per-context data
+storage.
+The storage is organized as a map for storing device code image. It uses
+different keys to address difference in SYCL objects ids between applications
+runs as well as the fact that the same kernel name can be used in different
+SYCL applications.
+
+The programs map's key consists of four components:
+ - device image id<sup>[1](#what-is-diid)</sup>,
+ - specialization constants values,
+ - device id<sup>[2](#what-is-did)</sup> this program is built for,
+ - build options id<sup>[3](#what-is-bopts)</sup>.
```

Comment:
The description does not handle situation when for two programs all the above match, but the source binaries they were compiled from differ. Can this be re-formulated to something like:
The cache is organized as a multi-map. Multi-map key includes \<those four fields above\> and source binary size. Matching keys don't guarantee item match, so for each item with matching key SYCL runtime will additionally compare the source binaries, which are kept as a part of the cached item object.

Reply:
Added description of hash collision case in the paragraph